[Switch] Create SwitchUnstyled and useSwitch

michaldudak opened the pull request:
This is an implementation of the SwitchUnstyled component and the useSwitch hook.

## Summary
* `useSwitch` is a low-level utility for both internal and external use when maximum customizability is required. It provides an object to be used as `input` props
* `SwitchUnstyled` creates a structure of components and uses `useSwitch` internally. `Root`, `Thumb` and `Input` components are overridable via `components` prop (and `component`, being equivalent to `components.Root`).
* The original Switch component used quite a deep hierarchy: Switch -> SwitchBase -> ButtonBase -> input. The new unstyled one (and, in consequence the Switch) does not use neither ButtonBase nor SwitchBase anymore. Part of the logic was moved to useSwitch (handling of focusVisible) and part was moved to a new `useTouchRipple` hook.
* Initially I indented to implment better type checking of `componentsProps`, but after encountering problems with PropTypes, I created a separate issue for this: https://github.com/mui-org/material-ui/issues/26810
* A new testing utility has been created: `describeConformanceUnstyled`. It executes the existing tests that apply to unstyled components + a few new ones (testing `components` and `componentsProps` props)

Preview: https://deploy-preview-26688--material-ui.netlify.app/components/switches/#unstyled-switches

## To do

- [x] tests
- [x] documentation
- [x] demos
- [ ]  minimal styles for SwitchUnstyled -> will be done in a separate PR if needed. For now, the unstyled component is truly unstyled.

mnajdova commented on packages/material-ui-unstyled/src/SwitchUnstyled/SwitchUnstyled.tsx:
```diff
@@ -0,0 +1,81 @@
+  };
+}
+
+const minimalRootStyles: CSSProperties = {
```
```suggestion
const baseRootStyles: CSSProperties = {
```
Or maybe even:
```suggestion
const rootStyles: CSSProperties = {
```

As we are already in an `Unstyled*` component.

oliviertassinari commented:
Great initiative to open a draft so we can discuss on the changes early. The first thing that got my attention is regarding the **constraints**. The existing components aren't using the new unstyled ones (yet?). What's the plan regarding this aspect?

I can think of two different approaches:

1. Unstyled button and unstyled switches are isolated from the existing components. Pros: no risk of integration that fail, more guidance by extracting what works, faster?
2. We start from scratch, we work on the API regardless of what already exists, and try to make it fit as a second step. Pros: reduce upfront constraints that allows to rethink it, higher quality? 

It seems that we are going with option 2. 

In the two cases, one big challenge is which level we extract/abstract. It has to be enough abstracted to make it valuable for developers (solve problems for them, otherwise they will write their own hook). It also has to be low level enough to allow us to implement a second design specification

oliviertassinari commented on docs/src/pages/components/switches/UseSwitches.js:
```diff
@@ -0,0 +1,84 @@
+});
+
+function FancySwitch(props) {
+  const { getInputProps, isChecked, isDisabled } = useSwitch(props);
```
It can be valuable for merging extra props for the developers, e.g. callback can be a pain.

michaldudak replied:
I followed the convention I found in `useAutocomplete`. And yes, it makes it easier to merge props.

mnajdova replied:
My point was, the hook is already a function of the props, and `getInputProps` doesn’t receive any args. Maybe it makes sense for more complex components.

oliviertassinari replied:
Using your source, Kent details it a bit in https://kentcdodds.com/blog/introducing-downshift-for-react#prop-getters

mnajdova replied:
It makes sense now that the `getInputProps` receives some args :) Resolving it.

michaldudak commented on docs/src/pages/components/switches/UseSwitches.js:
```diff
@@ -0,0 +1,84 @@
+
+  return (
+    <FancySwitchElement
+      className={clsx({ disabled: isDisabled, checked: isChecked })}
```
I actually didn't use MUI-specific classes on purpose, to show that developers are free to use whatever names they choose and it doesn't have to be anything that's already used across the core components.

oliviertassinari replied:
> show that developers are free to use whatever names they choose

But is this true? `.disabled` is hardcoded. I don't see the issue with using `.Mui-disabled`. The purpose of the Mui prefix was to avoid CSS conflicts with existing styles on the page. This class name is never styled directly.

mnajdova replied:
I didn't noticed it's a demo too, sorry :)

michaldudak commented on packages/material-ui-unstyled/src/ButtonUnstyled/ButtonUnstyled.tsx:
```diff
@@ -0,0 +1,260 @@
+  type?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
+  href?: string;
+  tabIndex?: number | string;
+  [otherProp: string]: unknown;
```
Actually, it's pretty easy to get rid of that by either:
1. having `ButtonUnstyled` props of type `ButtonUnstyledProps<TRoot> & ComponentPropsWithRef<TRoot>` (+ perhaps some `Omit` if necessary) - what would surface all the props of the `Root` component to the host, or
2. do not forward props from `ButtonUnstyled` to its `components.Root` and move the `role` to `componentsProps.root`.

Option 1 may be simpler to use, will create two places where you could pass a prop to the `Root` component - the host and `componentsProps.root`. Option 2 would make the API cleaner. I feel I have a slight preference towards option 2, but I'm open to suggestions.

michaldudak commented:
> Great initiative to open a draft so we can discuss on the changes early. The first thing that got my attention is regarding the **constraints**. The existing components aren't using the new unstyled ones (yet?). What's the plan regarding this aspect?

My next goal will be changing the Core `Switch` to use this implementation.

> I can think of two different approaches:
> 
> 1. Unstyled button and unstyled switches are isolated from the existing components. Pros: no risk of integration that fail, more guidance by extracting what works, faster?
> 2. We start from scratch, we work on the API regardless of what already exists, and try to make it fit as a second step. Pros: reduce upfront constraints that allows to rethink it, higher quality?
> 
> It seems that we are going with option 2.

I chose to build the components from scratch using the existing functionality as reference. Of course I copied a lot of code from the existing components. I find it easier to start with a blank slate and add things when I see they are necessary, than start from a full-blown component and remove things that are not needed. Additionally, I wanted to build these components in TypeScript from the start.

> In the two cases, one big challenge is which level we extract/abstract. It has to be enough abstracted to make it valuable for developers (solve problems for them, otherwise they will write their own hook). It also has to be low level enough to allow us to implement a second design specification

I agree - it's hard to draw a line properly. For sure, all the accessibility features and nothing Material-specific should be in unstyled. 
The sooner we start implementing the second theme, the sooner we will validate the decisions made to extracts the unstyled components. Then, we may revisit the unstyled API if necessary.

michaldudak commented on packages/material-ui-unstyled/src/SwitchUnstyled/SwitchUnstyled.tsx:
```diff
@@ -0,0 +1,81 @@
+};
+
+const SwitchUnstyled = function SwitchUnstyled<
+  TRoot extends ElementType = 'span',
```
Besides having typed `componentsProps`, I wanted to use the type system to enforce constraints on the inner components. In case of SwitchUnstyled, this means for example having `TRoot` with `children` prop. However, setting `TRoot extends ElementType<{children?: ReactNode}>` doesn't do the trick. I'll investigate this deeper and would appreciate anyone's (@eps1lon ?) ideas.

eps1lon replied:
> However, setting TRoot extends ElementType<{children?: ReactNode}> doesn't do the trick. I'll investigate this deeper and would appreciate anyone's (@eps1lon ?) ideas.

When you say it doesn't work do you mean that I could pass an element type that doesn't implement `children`? What happens if you constrain it further to `TRoot extends ElementType<children: JSX.Element>`? 

Or try `React.JSXElementConstructor` instead

michaldudak replied:
With `TRoot extends ElementType<{ children?: ReactNode }>` TypeScript gives up where the TRoot is instantiated saying "Expression produces a union type that is too complex to represent". I'm going to spend some time on this today so hopefully I'll find a solution.

eps1lon replied:
Yeah, just be careful not to spend too much time on it. These issues are very likely to turn out as not supported in TypeScript.

mnajdova commented on packages/material-ui-unstyled/src/ButtonUnstyled/ButtonUnstyled.tsx:
```diff
@@ -0,0 +1,252 @@
+
+  return (
+    <ButtonRoot
+      className={clsx(classes.root, className)}
```
We need go also merge here the className coming from `componentsProps.root`

oliviertassinari commented:
> My next goal will be changing the Core Switch to use this implementation.

@michaldudak This sounds like a great next step. I'm not even sure we will need to write dedicated test cases, as we could rely on the end-to-end tests with the integration of Material Design.

mnajdova commented:
> I'm not even sure we will need to write dedicated test cases, as we could rely on the end-to-end tests with the integration of Material Design (we have done mostly that so far for the unstyled components).

Agree, ideally we should not need to change any tests :)

michaldudak commented on docs/src/pages/components/switches/switches.md:
```diff
@@ -57,6 +57,14 @@ Here are some examples of customizing the component. You can learn more about th
 
+## Unstyled switches
+
+{{"demo": "pages/components/switches/UnstyledSwitches.js"}}
```
:) That was the goal - to show that you can easily create something much different than the original design. But yes, I agree the usefulness beyond that is minimal

michaldudak replied:
Demos updated

mnajdova commented on docs/src/pages/components/switches/UnstyledSwitches.tsx:
```diff
@@ -0,0 +1,85 @@
+      <SwitchUnstyled
+        checked={checked}
+        onChange={() => setChecked((c) => !c)}
+        components={{ Root }}
```
`component` should be complementary, both API should exists together.

oliviertassinari replied:
@michaldudak I see your point. I didn't realize it was done on purpose. In the community, another popular name of the `component` prop is `as`. There is a tension here. For components with one dom node, components are overkill. For components with multiple host dom nodes, a `component` prop has limited value.

I personally think that this is similar to the tension between `className` and `classes`, we ended always adding the two for consistency.

michaldudak replied:
OK, so:

For components with multiple slots we'll use `components` and `componentsProps`. Should there also be a `component` prop here (equivalent to `components.Root`)? Ref will be forwarded to the root element.
Do you have any thoughts about prop forwarding? If all unknown props will be forwarded to the root component, we'll have two ways of passing props to it (by placing them at the host and using `componentsProps.root`).

For components where just the root is customizable, we'll use the `component` prop and forward props from the host to the root. There will be no `componentProps` prop. Ref will be forwarded to the root element.

Let me know if I got this right.

oliviertassinari replied:
I would vote for always having the 3 props (component, components, componentsProps), to minimize surprises for the developers. 

For props forwarding, I personally don't have any specific thoughts other than root being forwarded to the root is what we have been aiming for, so far.

mnajdova replied:
+1 for always having all three props.

michaldudak commented on packages/material-ui-unstyled/src/SwitchUnstyled/useSwitch.ts:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    setCheckedState(event.target.checked);
+    onChange?.(event);
```
I'm trying to update these whenever I encounter them.

oliviertassinari replied:
Babel transpile them [a bit differently](https://babeljs.io/repl#?browsers=ie%2011%2C%20edge%20%3E%3D%2014%2C%20firefox%20%3E%3D%2052%2C%20chrome%20%3E%3D%2049%2C%20safari%20%3E%3D%2010%2C%20node%208.0%0A&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=PYOwwgFghiDmCmB-AdACngN3iALgSgG4g&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=env%2Creact&prettier=false&targets=&version=7.14.4&externalPlugins=). @eps1lon any preferences on this topic?

eps1lon replied:
> onChange?.(event);

:+1:  That's describing better what we want conceptually.

> if (onChange) onChange

:-1: This is not reasonable since it relies on JS implicit type conversions.

michaldudak commented on packages/material-ui-unstyled/src/SwitchUnstyled/SwitchUnstyled.tsx:
```diff
@@ -0,0 +1,638 @@
+      ]),
+      PropTypes.func,
+    ]),
+    Thumb: PropTypes.oneOfType([
+      PropTypes.oneOf([
+        'a',
+        'abbr',
+        'address',
+        'animate',
+        'animateMotion',
+        'animateTransform',
```
I haven't figured out (yet) how to use the type system to restrict the elements/components applicable to be placed in a slot. I still have to work on this.

oliviertassinari replied:
You would have to dig into this file, maybe here https://github.com/mui-org/material-ui/blob/ca001affdeba0a7adb4594bd16c0f6e3c5b07d72/scripts/generateProptypes.ts#L171

Sebastian is owning this logic.

eps1lon replied:
It should create a `PropTypes.elementType` but that might be incredibly tricky to determine with the compiler API considering we don't just allow `React.ElementType` but `T extends React.ElementType`.

I don't think it's useful to constrain these element types at this point considering we never do this in other parts of the codebase.

michaldudak replied:
I've spend good few hours on this, but couldn't find a sensible way to solve this. I propose we leave it like that for now and possibly improve in the future. Technically, these prop types are correct (-ish - as Root for example needs a non-empty element, so not all tags apply), even though they look quite ugly.

michaldudak commented on docs/src/pages/components/switches/UseSwitches.js:
```diff
@@ -0,0 +1,84 @@
+  );
+}
+
+export default function UseSwitch() {
```
Yes, now that I implemented the `Switch` using `useSwitch` and have proof that the API is decent, I'll focus more on updating the docs and demos.

michaldudak commented:
I implemented the Switch component using the `useSwitch` hook. Looks just like the original and the unit tests pass. Now I'll focus more on making the rest of the tests pass and to improve the docs.

Inspired by https://github.com/mui-org/material-ui/pull/26303 I decided to remove the ButtonBase from the render tree. FocusVisible support is provided by `useSwitch` and I moved handling the ripple effect to the new `useTouchRipple` hook. This way, I don't need to include ButtonUnstyled in this PR (I'll move it to another one).

oliviertassinari commented on packages/material-ui-unstyled/src/SwitchUnstyled/switchUnstyledClasses.ts:
```diff
@@ -0,0 +1,30 @@
+  return generateUtilityClass('SwitchUnstyled', slot);
+}
+
+const switchUnstyledClasses: SwitchUnstyledClasses = generateUtilityClasses('SwitchUnstyled', [
```
Are we sure about this Unstyled suffix? In practice, the class names will be used to style, so I don't really see the use case. The same classes will also be used by the default styled components.

```suggestion
const switchUnstyledClasses: SwitchUnstyledClasses = generateUtilityClasses('MuiSwitch', [
```

oliviertassinari commented on packages/material-ui-unstyled/src/SwitchUnstyled/SwitchUnstyled.tsx:
```diff
@@ -0,0 +1,177 @@
+    <Root
+      ref={ref}
+      {...rootProps}
+      className={clsx(classes.root, stateClasses, rootProps?.className)}
```
So, no support for the `classes` prop?

michaldudak replied:
Nope. We discussed it with @mnajdova. We already have `componentsProps` extension point and I feel that creating yet another one (that could potentially clash with `componentsProps.x.className`) won't help.

oliviertassinari replied:
Ok, Interesting. Let's see how this goes.

oliviertassinari commented on packages/material-ui-unstyled/src/SwitchUnstyled/useSwitch.ts:
```diff
@@ -0,0 +1,136 @@
+  const [checked, setCheckedState] = useControlled({
+    controlled: checkedProp,
+    default: Boolean(defaultChecked),
+    name: 'SwitchUnstyled',
```
In the autocomplete, we make this name configurable, likely overkill for now.

michaldudak replied:
Let's add in in the future if we see it's actually needed. For now, I'm considering having `Switch` or `MuiSwitch` instead of `SwitchUnstyled`. What do you think?

oliviertassinari replied:
No really preferences, I personally like the name of the module, what used in the Autocomplete:
```suggestion
    name: 'useSwitch',
```

michaldudak replied:
I'm leaning towards `Switch` here. This name is visible to the developer when there's a change between uncontrolled and controlled component. Developers don't need to know that the Swich uses `useSwitch` internally, so having just `Switch` here would improve the meaning of the message IMO.

oliviertassinari commented on packages/material-ui/src/Switch/Switch.js:
```diff
@@ -110,23 +86,56 @@ const SwitchSwitchBase = styled(SwitchBase, {
       duration: theme.transitions.duration.shortest,
     }),
+    // from SwitchBase
+    /* Styles applied to the root element. */
```
No need to keep these comments, we want to remove them all.

```suggestion
```

oliviertassinari reviewed (commented):
It seems that we are close 👍

michaldudak commented on packages/material-ui/src/Switch/Switch.js:
```diff
@@ -30,64 +40,30 @@ const useUtilityClasses = (styleProps) => {
   const composedClasses = composeClasses(slots, getSwitchUtilityClass, classes);
```
How would developers add custom classes to the thumb or track, then?

oliviertassinari replied:
`composeClasses` is already merging the slots with the `classes.

michaldudak replied:
🤦 indeed 😉

michaldudak commented on packages/material-ui-unstyled/src/SwitchUnstyled/useSwitch.ts:
```diff
@@ -0,0 +1,136 @@
+  const handleRefChange = useForkRef(focusVisibleRef, inputRef);
+
+  return {
+    getInputProps: (otherProps: object = {}) => ({
```
If it doesn't matter what's the element, I'd be happy with `getRootProps` but in this case it must be an input (or something that forwards these props to the input, so I think `getInputProps` is a better option here.

mnajdova replied:
I'd say we should use the same word for the slot everywhere, example inside `components`,  `componentsProps`, `classes` if we support it, and `getSlotProps` :) So I would vote for `getRootProps` too.

michaldudak replied:
But these are not the root slot properties

michaldudak commented on docs/src/pages/components/switches/UseSwitchesCustom.tsx:
```diff
@@ -0,0 +1,99 @@
+export default function UseSwitchesCustom() {
+  return (
+    <div>
+      <MaterialSwitch defaultChecked />
```
This demo shows the Material Design switch (as in https://next.material-ui.com/components/switches/#customized-switches). As for "MUI", in the future this may be a broader term than just Material, no? :)

oliviertassinari replied:
I believe this demo is:

<img width="252" alt="Capture d’écran 2021-06-21 à 15 21 25" src="https://user-images.githubusercontent.com/3165635/122768930-5f48ad00-d2a4-11eb-9279-5ee1de2d29b9.png">

from the rebranding of the company. MUI as a potential name for the project, to reduce the emphasis on Material Design.

mnajdova commented on docs/src/pages/components/switches/UseSwitchesCustom.tsx:
```diff
@@ -0,0 +1,99 @@
+import * as React from 'react';
+import clsx from 'clsx';
+import { styled } from '@material-ui/system';
+import { useSwitch, UseSwitchProps } from '@material-ui/unstyled/SwitchUnstyled';
```
Yeah, I should update the unstyled examples :) They shouldn't have any import from the `@material-ui/core`.

michaldudak commented on docs/src/pages/components/switches/switches.md:
````diff
@@ -57,6 +57,33 @@ Here are some examples of customizing the component. You can learn more about th
+For the ultimate customizability, a `useSwitch` hook is available. It accepts almost the same options as the SwitchUnstyled component minus the `component`, `components` and `componentsProps` props.
+The Switch component is [built using this hook](https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Switch/Switch.js).
+
+```jsx
+import { useSwitch } from '@material-ui/unstyled/SwitchUnstyled';
+```
````
My personal preference would be to use named exports everywhere (I agree with https://basarat.gitbook.io/typescript/main-1/defaultisbad) but I don't intend to start a revolution :)

oliviertassinari replied:
We had a long debate about this problem in https://github.com/mui-org/material-ui/issues/21862, with no real outcome

oliviertassinari replied:
> My personal preference would be to use named exports everywhere

Then you would probably personally prefer:

```jsx
import { useSwitch } from '@material-ui/unstyled';
```

oliviertassinari commented on docs/src/pages/components/switches/switches.md:
```diff
@@ -57,6 +57,47 @@ Here are some examples of customizing the component. You can learn more about th
+
+{{"demo": "pages/components/switches/UnstyledSwitchesMaterial.js"}}
+
+### useSwitch hook
```
I like that we have demos with headless and unstyled API. It will allow us to look at the GA metrics and see which one gets more usage.

My gut intuition is that headless will be a bit more popular than the unstyled API in its current form. My reasoning:

- Developers coming from a different context would rather use the headless API to understand what's going on.
- Developers that want to go one layer down from MD can't use Unstyled directly because it's not the same component structure (because we didn't use Untyled in MD).

oliviertassinari replied:
> do you think it'll still be the case when we finally switch to use unstyled in MD?

I'm making wild bets here 😆.

Another interesting thing: I have heard multiple times design system leaders using our components with customizations that and they don't have the time to build great documentation, that they push the customization so they can work as much as possible with our documentation.

oliviertassinari commented on docs/src/pages/components/switches/UnstyledSwitches.tsx:
```diff
@@ -0,0 +1,88 @@
+        componentsProps={{ input: withDemoLabel, root: { as: 'div' } }}
+      />
+      <SwitchUnstyled
+        components={{ Root }}
```
We can do it now, and it's actually simpler to customize with the CSS selectors targeting nested elements.

```suggestion
        component={Root}
```

oliviertassinari commented on docs/src/pages/components/switches/switches.md:
```diff
@@ -57,6 +57,47 @@ Here are some examples of customizing the component. You can learn more about th
+
+{{"demo": "pages/components/switches/UnstyledSwitches.js"}}
+
+### Recreation of Material-UI's Switch
+
+{{"demo": "pages/components/switches/UnstyledSwitchesMaterial.js"}}
```
Basically, these MD demos seem mostly here to have a constraint. Once, that will ultimately come from the MD and second theme components.

oliviertassinari reviewed (commented):
- I personally don't buy into the design of the Unstyled switches customization demo. It feels hard to differentiate the on to off states. We can involve @daniloleal to help, we started the thread in:
https://github.com/mui-org/material-ui/pull/26688#discussion_r650594489.
- I'm OK starting without the constraint of the MD for unstyled. But I think that we shouldn't leave the alpha state for the unstyled package until we have upstream dependencies (MD + second themes). The sooner the better.
- For "minimal styles for SwitchUnstyled". I would personally vote against it in the current state of the information available. I see the appeal, but I think that it's counterproductive, per the constraints it will set. I also have no doubt developers will ask it becomes a common pain. So we could delay.
- We have about 50 components to support in the long term, how do you see the approach scaling?
- Outside of the comments I have left. I can't see anything else. It looks solid. Great work!

For context, I think that most (2/3 my guess) of the value of these components in the coming 12 months is going to be with @siriwatknp and sharing the same logic between two design systems. The market of unstyled raw has yet to show [a prime time](https://npm-stat.com/charts.html?package=reakit&package=%40react-aria%2Fselect&package=%40reach%2Fdialog&package=react-dom&from=2020-06-25&to=2021-06-25): I think that it needs the proof that two popular and different design systems depend on it (and hopefully more) to start selling to the community. So I don't think that we should lose this objective in mind: **unstyled as a means for a second theme**. For the advanced components, I think that it's a different story, I think that a lot more valuable, https://github.com/iamhosseindhv/notistack/issues/184 should be enough to prove it.

michaldudak commented on docs/src/modules/utils/parseTest.ts:
```diff
@@ -34,10 +34,7 @@ function findConformanceDescriptor(file: babel.ParseResult): babel.types.ObjectE
-        t.isIdentifier(callee) &&
-        (callee.name === 'describeConformance' || callee.name === 'describeConformanceV5')
-      ) {
+      if (t.isIdentifier(callee) && callee.name.startsWith('describeConformance')) {
```
Yes, I'm aware of that. The reason I changed this is that I added another set of tests, `describeConformanceUnstyled`, tailored specifically for unstyled components.

michaldudak commented on docs/src/pages/components/switches/UnstyledSwitches.tsx:
```diff
@@ -0,0 +1,88 @@
+        checked={checked}
+        onChange={() => setChecked((prevChecked) => !prevChecked)}
+        components={{ Root }}
+        componentsProps={{ input: withDemoLabel, root: { as: 'div' } }}
```
OK, sure, I'll make it simpler.

michaldudak reviewed (commented):
> 1. I personally don't buy into the design of the Unstyled switches customization demo. It feels hard to differentiate the on to off states. We can involve @daniloleal to help, we started the thread in:
>    [[SwitchUnstyled] Create SwitchUnstyled and useSwitch #26688 (comment)](https://github.com/mui-org/material-ui/pull/26688#discussion_r650594489).
> 
> https://deploy-preview-26688--material-ui.netlify.app/components/switches/#unstyled-switches
> 
> <img alt="Capture d’écran 2021-06-26 à 23 37 58" width="162" src="https://user-images.githubusercontent.com/3165635/123526352-8f92b000-d6d7-11eb-93b7-5106a8766577.png">

Sure, if @danilo-leal can design something prettier, I'm happy to restyle it.

> 2. I'm OK starting without the constraint of the MD for unstyled. But I think that we shouldn't leave the alpha state for the unstyled package until we have upstream dependencies (MD + second themes). The sooner MD depends on unstyled the better. IMHO the more a component is designed focused, the more it's in our interest to use unstyled inside MD from the get-go.

As discussed internally, we can push unstyled as a base for MD in v5.1

> 4. We have about 50 components to support in the long term, how do you see the approach scaling?

The sooner these components get used as a base of MD/2nd theme, the better - we need to avoid having two distinct codebases to maintain.

> 5. Outside of the comments I have left. I can't see anything else. It looks solid. Great work!

Good to hear, thanks!

> 6. For context, I think that most (2/3 my guess) of the value of these components in the coming 12 months is going to be with @siriwatknp and sharing the same logic between two design systems. The market of unstyled raw has yet to show [a prime time](https://npm-stat.com/charts.html?package=reakit&package=%40react-aria%2Fselect&package=%40reach%2Fdialog&package=react-dom&from=2020-06-25&to=2021-06-25) (nor that it can be well monetized). I think that for unstyled to really grow it needs proof that two popular and different design systems depend on it (and hopefully more) to start selling to the community. So I don't think that we should lose this objective in mind: **unstyled to get an efficient/effective second theme**. For more advanced components than a Switch, I think that it's a different story, I think that a lot more valuable, [Notistack without Material-UI? [POLL] 👍👎 iamhosseindhv/notistack#184](https://github.com/iamhosseindhv/notistack/issues/184) should be enough to prove it.

Surely it'll be interesting to see @siriwatknp (and community) use these components/hooks. I'm eager to see the feedback. We'll iterate on the API design if needed.

danilo-leal commented:
> Sure, if @danilo-leal can design something prettier, I'm happy to restyle it.

I've made small tweaks to polish it. There's not much I guess for a simple switch 😅 Take a look and tell me what do you think.
https://codesandbox.io/s/useswitchesbasic-material-demo-forked-9zijz?file=/demo.js

michaldudak commented:
Clearly it's visible now when the switch is checked and unchecked but IMO right now the distinction between normal unckecked and disabled unchecked states is not enough.

danilo-leal commented:
> Clearly it's visible now when the switch is checked and unchecked but IMO right now the distinction between normal unckecked and disabled unchecked states is not enough.

Thanks, I think now it's indeed better, please take another peek at it. 
https://codesandbox.io/s/useswitchesbasic-material-demo-forked-9zijz?file=/demo.js

michaldudak commented:
That's better, thanks. @oliviertassinari, what would you say?

michaldudak commented on packages/material-ui-system/src/createStyled.js:
```diff
@@ -52,8 +52,15 @@ const variantsResolver = (props, styles, theme, name) => {
+    prop !== 'theme' &&
+    prop !== 'sx' &&
+    prop !== 'as' &&
+    prop !== 'componentState'
```
Actually, can we follow emotion's approach and do not forward any custom props if the styled element is an HTML element? (cc @mnajdova)

mnajdova replied:
> Actually, can we follow emotion's approach and do not forward any custom props if the styled element is an HTML element? (cc @mnajdova)

Yeah, we could.

---

A bit background so that we can decide on `styleProps` vs `componentState`. When I initially introduced this prop, it was meant to only include props which are used in the styles. Then we saw the need for forwarding here more props (for example custom props for styling - maybe new color prop?). Currently the `styleProps` are implemented as all props + default values + state.

For the `styleProps`, the intention in the name looks better for me. If we name them `componentState`, then there will be question on, why some state is sent as a first class prop, and some is only in the `componentState`. Although the components are `unstyled` they should offer some injection point for applying styles, hence the name `styleProps`.

Anyway, the most important thing is that, we need a bag for all props so that we don't have to check every prop and decide whether it needs to be propagated to the DOM element or not.

oliviertassinari replied:
> Actually, can we follow emotion's approach and do not forward any custom props if the styled element is an HTML element?

@michaldudak Do you mean using `@emotion/is-prop-valid`? AFAIK, we are no using it. So far, we intercept the props we don't want to leak, one by one. It has a nice side effect to forces us to be more aware of how props flow. It might also have negative performance implications to use it. 

It seems that sc and em use it by default, while we don't. See the DOM output of: https://codesandbox.io/s/elegant-feather-5myf3?file=/src/App.js.

```jsx
import "./styles.css";
import styled1 from "styled-components";
import styled2 from "@emotion/styled";
import { styled as styled3 } from "@material-ui/system";

const Div1 = styled1("div")({});
const Div2 = styled2("div")({});
const Div3 = styled3("div")({});

export default function App() {
  return (
    <div className="App">
      <Div1 foo="foo">1</Div1>
      <Div2 foo="foo">2</Div2>
      <Div3 foo="foo">3</Div3>
    </div>
  );
}
```

<img width="278" alt="Capture d’écran 2021-06-29 à 12 57 45" src="https://user-images.githubusercontent.com/3165635/123786114-9ee14b80-d8d9-11eb-8c8e-ff616484bd34.png">

---

For the name `stateProps` could work too over `styleProps` and `componentState`.

michaldudak replied:
Yes, I meant `@emotion/is-prop-valid` (just for elements being HTML nodes). The performance hit should be neglible as they memoize the function call. But you're right that in most cases it shouldn't be necessary.

I think it's worth documenting that even though we are based on emotion, our behavior in this matter differs from theirs (BTW when I started working on this I used to spend some time and lost few hairs trying to figure out what's eating my `styleProps` prop... 😉).

As for the `styleProps`/`componentState` - after some thought I'm ok with passing all props and merging them with internal component state (although having event handlers there doesn't make much sense). 

Now, for the name - when I first saw it, I had hard time figuring out what `styleProps` are. Intuitively I thought it's a subset of props that's used for styling. First of all, they are technically not exactly props (they are host's props enriched with the computed state), so `stateProps` won't be my preference either. From the perspective of a receiving (slot) component, it's an object that describes the state of the host component. I suppose that in most cases it'll be indeed used to style the custom component, but it doesn't have to (nor should we care what it's used for). Perhaps `hostState` would suit you better, then?

michaldudak replied:
RFC created: https://github.com/mui-org/material-ui/issues/27127

oliviertassinari commented on packages/material-ui-unstyled/src/SwitchUnstyled/switchUnstyledClasses.ts:
```diff
@@ -0,0 +1,40 @@
+  input: string;
+  /** Class applied to the thumb element */
+  thumb: string;
+  /** Class applied to the root element if the switch is checked */
```
Because:

<img width="482" alt="Capture d’écran 2021-06-29 à 12 44 51" src="https://user-images.githubusercontent.com/3165635/123784525-d222db00-d8d7-11eb-995b-35658b993b4e.png">

michaldudak replied:
Sorry but I don't get it - it's still an ordinary CSS class, isn't it? A term "pseudo-class" has a specific meaning in CSS and if we call normal classes this way, I can't see anything but confusion.

oliviertassinari replied:
The class, e.g; `.Mui-checked` can't be styled directly with a CSS selector. If a developer does it, he would override all the others components. In the docs, we have used the term [pseudo-class](https://material-ui.com/customization/components/#pseudo-classes) to signal the difference in behavior. My reasoning was that it behaves differently, therefore it needs to have something different in the description. Happy with an alternative terminology description.

michaldudak replied:
Yet again naming things proves to be the [hardest part of programming](https://twitter.com/secretGeek/status/7269997868) :)

You're absolutely right that we should differentiate these classes from the ones that always apply. My concern is with the term pseudo-class and it's overloaded meaning. If a new developer (or even me for that matter) comes in and sees it, they might associate it with CSS pseudo-classes prefixed by a colon and start (unsuccessfully) authoring styles using the same pattern. Since these classes are applied _conditionally_, perhaps "conditional classes" would be a better term?

oliviertassinari replied:
First, to be clear, this thread is off-topic to the effort of the PR. I have started it as the context seems relevant.

The BEM convention has something close: the modifier, but we probably can't use it because readOnly is also a modifier but doesn't get an extra layer of CSS specificity. If we go back to why this exists, it's about allowing developers to customize the baseline style, without messing with the common state styles. "state classes" could be another option but maybe too generic. I like "conditional classes", it's unique, and quick close to the underlying problem. @mnajdova Should we open an RFC/PR to do the change in the codebase?

michaldudak replied:
RFC created: https://github.com/mui-org/material-ui/issues/27128

oliviertassinari commented:
@michaldudak It looks great 👌 

<img width="212" alt="Capture d’écran 2021-06-29 à 13 09 02" src="https://user-images.githubusercontent.com/3165635/123787503-309d8880-d8db-11eb-8cc8-5fcadf85acba.png">

mnajdova reviewed (commented):
Looks pretty close to be ready for merge :)

michaldudak commented on docs/src/pages/components/switches/UseSwitchesCustom.js:
```diff
@@ -0,0 +1,95 @@
+  const { getInputProps, checked, disabled, focusVisible } = useSwitch(props);
+
+  const stateClasses = {
+    checked,
```
👍🏻 I did this in UseSwitchesBasic.

michaldudak commented:
@oliviertassinari, @mnajdova I addressed all the comments you made. Could you please take a final look at the PR and see if there's anything outstanding from your point of view?

oliviertassinari commented on docs/src/pages/components/switches/UseSwitchesMaterial.js:
```diff
@@ -0,0 +1,413 @@
+        );
+      }
+    }, [enableTouchRipple]);
+  }
```
@mnajdova It depends on the objective of this demo. AFAIK, it's not mean to stay in the long run, but to replace the existing material design component. We have it in the docs, to help spot regressions

oliviertassinari commented on packages/material-ui-unstyled/src/SwitchUnstyled/SwitchUnstyled.tsx:
```diff
@@ -0,0 +1,211 @@
+  } = props;
+
+  const Root: React.ElementType = component ?? components.Root ?? 'span';
+  const rootProps: any = { ...getForwardableProps(Root, otherProps), ...componentsProps.root };
```
Why is this not enough?

```suggestion
  const rootProps: any = { ...otherProps, ...componentsProps.root };
```

michaldudak replied:
If `Root` is a plain HTML element, I don't want to forward anything that doesn't apply to a HTML element.

oliviertassinari replied:
The developers can swallow the extra prop. We only check it with the "other" props on the root element, but not with the `componentsProps` slots.

I mean, I don't get why foo shouldn't warn when it does with the buttons:

```jsx
import * as React from "react";
import Button from "@material-ui/core/Button";
import SwitchUnstyled from "@material-ui/unstyled/SwitchUnstyled";

export default function UnstyledSwitches() {
  return (
    <div>
      <SwitchUnstyled foo />
      <Button foo />
      <button foo />
    </div>
  );
}
```

https://codesandbox.io/s/unstyledswitches-material-demo-forked-rsrd1?file=/demo.tsx

michaldudak replied:
My thinking was that forwarding props from the host to the root is less explicit than just setting them in `componentsProps.root` and therefore it may need some additional checking. If someone explicitly passes something in `componentsProps.*`, I assume they know what they are doing.

This has an inconvenience of being inconsistent, though. I'm not particularly attached to this piece of code and if you feel we should always forward everything, so be it.

oliviertassinari replied:
cc @mui-org/core for more perspectives on the matter.

Outside of this point the PR is good to go, from my side.

mnajdova replied:
I would forward all props, as it was done before. Let's not complicate the code for problems we haven't encounter so far.

mnajdova reviewed (approved):
One tiny last comment for improving the demo. Let's also resolve https://github.com/mui-org/material-ui/pull/26688#discussion_r663931154 before merging. Well done 🚀

michaldudak commented on docs/src/pages/components/switches/UnstyledSwitchesMaterial.js:
```diff
@@ -0,0 +1,434 @@
+  zIndex: 1,
+});
+
+const SwitchThumb = ({ isChecked, icon, checkedIcon, className }) => {
+  if (!isChecked && icon) {
+    return icon;
+  }
+
+  if (isChecked && checkedIcon) {
+    return checkedIcon;
+  }
+
+  return <DefaultSwitchThumb className={className} />;
+};
```
This isn't equivalent. `icon` and `checkedIcon` shouldn't be styled by us.

mnajdova replied:
What do you mean? They are not, they would be styled if we add the `className` prop no?

michaldudak replied:
In your suggestion, if `icon` or `checkedIcon` is provided, they would have the styles applied (boxShadow, backgroundColor, etc.), which is not what we want

khushalsrashtasoft commented:
![image](https://github.com/mui/material-ui/assets/110820073/01467f1f-63b8-4973-966c-d1be50dcdd11)

 use i am import {  Switch } from "@mui/material";

I have this issue. please fix me.
